### DISCUSSION TITLE
How to set aggregation weights

### DISCUSSION THREAD

**Discussion by wanderine** [NVIDIA/NVFlare#1911] [Q&A] [answered]
I would like to set the aggregation weights to be equal to the number of training samples for each client, I cannot see any example of this in the examples, only setting fix weights. I will run many trainings with different data for my clients, and I want the client weights to be set automatically, so I don't have to write it manually every time.

**Comment by chesterxgchen** [comment]
@ZiyueXu77  can you share the example where this is set ?

**Comment by ZiyueXu77** [comment, marked as answer]
With the default learner/aggregator, the aggr weights are set to the number of traning samples, by transfering the below MetaKey in the local learner ("MetaKey.NUM_STEPS_CURRENT_ROUND") in 
https://github.com/NVIDIA/NVFlare/blob/main/examples/hello-world/hello-pt/jobs/hello-pt/app/custom/cifar10trainer.py#L145

To clarify, the "aggregation_weights" in the config json is in fact an "additional weight", that can be further set as
https://github.com/NVIDIA/NVFlare/blob/main/examples/hello-world/hello-pt/jobs/hello-pt/app/config/config_fed_server.json#L25C1-L27C24

And the two will be multiplied in the server aggregator as
https://github.com/NVIDIA/NVFlare/blob/main/nvflare/app_common/aggregators/dxo_aggregator.py#L130
and
https://github.com/NVIDIA/NVFlare/blob/main/nvflare/app_common/aggregators/dxo_aggregator.py#L160

**Reply by magdalena-fuchs** [reply]
Is there a more general answer to this question in case one does want to compute custom weights dynamically (e.g. based on certain training metrics)? What is the standard recommended way to do this in NVFlare? I am assuming a custom aggregator will be necessary.

**Reply by ZiyueXu77** [reply]
Setting an "outgoing DXO" is a necessary step for local trainer, and by default the NUM_STEPS_CURRENT_ROUND will be used to weight the response from clients (e.g. https://github.com/NVIDIA/NVFlare/blob/main/examples/hello-world/hello-pt/jobs/hello-pt/app/custom/cifar10trainer.py#L145). Therefore, just set this key to the value of the custom weights computed will do the trick. Assuming you have a function to calculate weights `dyn_weight = weights_based_on_metric()`, just set the outgoing DXO as
`meta={MetaKey.NUM_STEPS_CURRENT_ROUND: weights_based_on_metric()`
